govcd/nsxt_distributed_firewall_test.go

```diff
@@ -0,0 +1,272 @@
+//go:build network || nsxt || functional || openapi || ALL
+// +build network nsxt functional openapi ALL
+
+package govcd
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	. "gopkg.in/check.v1"
+)
+
+// Test_NsxtDistributedFirewall creates a list of distributed firewall rules with randomized parameters
+func (vcd *TestVCD) Test_NsxtDistributedFirewallRules(check *C) {
+	skipNoNsxtConfiguration(vcd, check)
+	skipOpenApiEndpointTest(vcd, check, types.OpenApiPathVersion1_0_0+types.OpenApiEndpointEdgeGateways)
+
+	adminOrg, err := vcd.client.GetAdminOrgByName(vcd.config.VCD.Org)
+	check.Assert(adminOrg, NotNil)
+	check.Assert(err, IsNil)
+
+	nsxtExternalNetwork, err := GetExternalNetworkV2ByName(vcd.client, vcd.config.VCD.Nsxt.ExternalNetwork)
+	check.Assert(nsxtExternalNetwork, NotNil)
+	check.Assert(err, IsNil)
+
+	vdc, vdcGroup := test_CreateVdcGroup(check, adminOrg, vcd)
+	check.Assert(vdc, NotNil)
+	check.Assert(vdcGroup, NotNil)
+
+	_, err = vdcGroup.ActivateDfw()
+	check.Assert(err, IsNil)
+
+	// Get existing firewall rule configuration
+	fwRules, err := vdcGroup.GetDistributedFirewall()
+	check.Assert(err, IsNil)
+	check.Assert(fwRules.DistributedFirewallRuleContainer.Values, NotNil)
+
+	// Create some prerequisites and generate firewall rule configurations to feed them into config
+	randomizedFwRuleDefs, ipSet, secGroup := createDistributedFirewallDefinitions(check, vcd, vdcGroup.VdcGroup.Id)
+
+	// Add IP Set and to cleanup list
+	openApiEndpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointFirewallGroups + ipSet.NsxtFirewallGroup.ID
+	PrependToCleanupListOpenApi(ipSet.NsxtFirewallGroup.Name, check.TestName(), openApiEndpoint)
+	openApiEndpoint = types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointFirewallGroups + secGroup.NsxtFirewallGroup.ID
+	PrependToCleanupListOpenApi(secGroup.NsxtFirewallGroup.Name, check.TestName(), openApiEndpoint)
```

Comment:
I may be wrong, but isn't this done also at the end of `preCreateVdcGroupSecurityGroup` (which is called by the above `createDistributedFirewallDefinitions`)

Reply:
Totally right. I the difference was that in these functions I had `AddToCleanupListOpenApi` (which adds to the env of cleanup), but in reality I needed `PrependToCleanupListOpenApi` which adds to the front of cleanup list.

The reason it needs to be cleaned up early is that it must be cleaned up before VDC Group as VDC Group removal fails if there stil are IP sets and/or security groups in it.